### PR TITLE
Add special handling for /dev in userbinds

### LINKED ISSUE #1907: Option to bind full /dev with contain
With 3.0 have an option to skip the minimal dev if `--contain`/`--containall` is used.

For example:

  `singularity exec --bind-dev -c [...]` 

Will do the normal `/dev` bind mounting instead of the minimal `/dev`. This _should not_ override the configuration setting of:  `mount dev = minimal`

### PR DESCRIPTION
**Pull request by cclerget**
**Description of the Pull Request (PR):**

Fixes #1907

### CODE REVIEW

**Comment by jmstover** [comment]
@cclerget 
Per @gmkurtzer , this needs to do this:  If the `singularity.conf` has `mount dev = minimal` then a user _must not_ be able to do `-B /dev`

File: src/runtime/engines/singularity/container.go [hunk 1]
```diff
@@ -989,6 +989,21 @@ func (c *container) addUserbindsMount(system *mount.System) error {
...
+			// by example with --contain option
+			if c.engine.EngineConfig.File.MountDev == "yes" {
+				system.Points.RemoveByTag(mount.DevTag)
+
+				err := system.Points.AddBind(mount.DevTag, "/dev", dst, syscall.MS_BIND|syscall.MS_REC)
+				if err != nil {
+					return fmt.Errorf("unable to add dev to mount list: %s", err)
+				}
+			} else {
+				sylog.Verbosef("Skipping /dev bind mount, disallowed by configuration")
```
Comment (GodloveD) [inline src/runtime/engines/singularity/container.go]: I think this should be a `sylog.Warningf` not just a verbose message.

**Review by GodloveD — CHANGES_REQUESTED** [review]
I think this looks great with the one small change I've suggested.

**Comment by cclerget** [comment]
@jmstover Yes this line https://github.com/singularityware/singularity/blob/f604c3055e472ca2eb40c46da2aa552ff3cb4add/src/runtime/engines/singularity/container.go#L995 check that and display warning message as suggested by @GodloveD if `mount dev` is not equal to `yes`

**Review by GodloveD — DISMISSED** [review]
lgtm

File: src/runtime/engines/singularity/container.go [hunk 2]
```diff
@@ -1001,27 +1032,42 @@ func (c *container) addUserbindsMount(system *mount.System) error {
...
-				err := system.Points.AddBind(mount.DevTag, "/dev", dst, syscall.MS_BIND|syscall.MS_REC)
-				if err != nil {
-					return fmt.Errorf("unable to add dev to mount list: %s", err)
+		// special case for /dev mount to override default mount behaviour
+		// with --contain option or 'mount dev = minimal'
+		if strings.HasPrefix(src, "/dev") {
+			if c.engine.EngineConfig.File.MountDev == "minimal" || c.engine.EngineConfig.GetContain() {
+				if strings.HasPrefix(src, "/dev/shm") || strings.HasPrefix(src, "/dev/mqueue") {
+					sylog.Warningf("Skipping %s bind mount: not authorized", src)
+				}
```
Comment (DrDaveD) [inline src/runtime/engines/singularity/container.go]: A fresh /dev/shm is already mounted so a request to bind that should get rejected as a duplicate bind, right?

As for /dev/mqueue, I don't see why someone would want to do that if they want to restrict the container, but if they ask for it, why should singularity prevent it?

Comment (cclerget) [reply src/runtime/engines/singularity/container.go]: The main reason is that these mount points are world writable and don't want to deal with nasty symlinks for security reasons, so Singularity just prevent bind and display a `not authorized` message

Comment (DrDaveD) [reply src/runtime/engines/singularity/container.go]: Oh I see, you're checking the prefix of the path, not only an exact match, I hadn't noticed.  Maybe those paths should end in slash though, so they don't prevent mounting the device directories themselves.

**Comment by cclerget** [comment]
I added missing mount of /dev/mqueue when IPC namespace and `--contain` are specified. This should be ready for review @GodloveD

**Review by jmstover — APPROVED** [review]

**Review by GodloveD — APPROVED** [review]
This lgtm as soon as conflicts are resolved.

**Pull request merged**